sys/mips/mips/trap.c

```diff
@@ -1162,6 +1205,19 @@ trap(struct trapframe *trapframe)
 		break;
 #ifdef CPU_CHERI
 	case T_C2E:
+		ftype = 0;
+
+		if (!c2e_fixup_fault(trapframe, true, p->p_vmspace, &ftype)) {
+			if (ftype != 0) {
+				if (KERNLAND(trapframe->badvaddr))
+					goto kernel_fault;
+				else
+					goto dofault;
+			}
+		} else {
+			return (trapframe->pc);
+		}
```

Comment:
These would make more sense to me if you flipped the order and hoisted the slow path out below, ie:

```
if (c2_fixup_fault(...))
	return (trapframe->pc);

if (ftype != 0) {
	...
}
```